### PR TITLE
Base logic pkg

### PR DESCRIPTION
**Pull request by mfdeveloper508**
ISO 8583 data fields are slightly different based on the network.
We would like to use a JSON configuration file for parsing the data elements rather than hard coding the logic for every use case.
we need to implement following items:
- Specifying format of configuration file
- Create dynamic data types with form configuration file
- Encoding/Decoding/Bytes logic by category of types
- Validation feature of dynamic data type
- Test Cases

### CODE REVIEW

File: pkg/lib/elements.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	}
+	return &DataElements{
+		Elements: make(map[int]*Element),
+		Spec:     spec,
+	}, nil
+}
+
+// data elements of the iso8583 message
+type DataElements struct {
+	Elements map[int]*Element
```
Comment (wadearnold) [inline pkg/lib/elements.go:40]: comment exported fields

File: pkg/lib/message.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+		},
+		Elements: _elements,
+		Spec:     spec,
+	}, nil
+}
+
+// message instance
+// isoMessage is structure for ISO 8583 message encode and decode
+type isoMessage struct {
+	Mti          *Element             `xml:"mti,omitempty" json:"mti,omitempty" yaml:"mti,omitempty"`
```
Comment (wadearnold) [inline pkg/lib/message.go:52]: @adamdecaf this is a MTI and acrynom. Should it all be capitalized?

File: pkg/utils/bcd.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+// license that can be found in the LICENSE file.
+
+package utils
+
+import (
+	"github.com/yerden/go-util/bcd"
+)
+
+var standard = &bcd.BCD{
+	Map: map[byte]byte{
```
Comment (wadearnold) [inline pkg/utils/bcd.go:12]: We have BTI inline with the library but we don't do it with EBCIDIC. Should these but part of this library or packages?

File: pkg/utils/version1987.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
+// Copyright 2020 The Moov Authors
+// Use of this source code is governed by an Apache License
+// license that can be found in the LICENSE file.
+
+package utils
+
+var (
+	ISO8583DataElementsVer1987 = Specification{
```
Comment (wadearnold) [inline pkg/utils/version1987.go:8]: @mfdeveloper508 this is awesome!

File: pkg/utils/charset.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+		res, err := transformEnconding(reader, charmap.Windows1252.NewEncoder())
+		return res, err
+	}
+	return input, nil
+}
+
+// BitMapArrayToHex converts a iso8583 bit array into a hex string
+func BitMapArrayToHex(arr []int64) (string, error) {
+	length := len(arr)
+	m := make(map[float64]string)
```
[the remark this answers is not in the document]
Comment (adamdecaf) [reply pkg/utils/charset.go:42]: Can we add tests for these conversions?

File: pkg/utils/types.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+func (s Attribute) ElementType() (*ElementType, error) {
+	attribute := s.Describe
+	for _, indicate := range digitIndicates {
+		if strings.Contains(attribute, indicate) {
+			var format string
+			if strings.Contains(attribute, formatIndicate) {
+				splits := strings.Split(attribute, formatIndicate)
+				attribute = splits[0]
+				format = strings.TrimSpace(splits[len(splits)-1])
```
Comment (adamdecaf) [inline pkg/utils/types.go:40]: We need a length check here. `strings.Split` will always return at least one item, but could return only one.

File: pkg/utils/types.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+			}
+
+			isFixed := !strings.Contains(indicate, variableIndicate)
+			splits := strings.Split(attribute, indicate)
+			_size, err := strconv.Atoi(strings.TrimSpace(splits[len(splits)-1]))
```
Comment (adamdecaf) [inline pkg/utils/types.go:45]: Same need for a length check.

File: go.mod [hunk 8]
```diff
@@ -3,6 +3,8 @@ module github.com/moov-io/iso8583
 go 1.14
 
 require (
+	github.com/Intermernet/ebcdic v0.0.0-20180129095110-0456a2287e54
```
[the remark this answers is not in the document]
Comment (adamdecaf) [reply go.mod:6]: Wouldn't https://github.com/gdamore/encoding/blob/master/ebcdic.go work for this? We use [that in ImageCashLeter already](https://github.com/moov-io/imagecashletter/blob/master/go.mod). 

cc @mfdeveloper508

File: pkg/lib/message.go (hunk 2 above)
Comment (adamdecaf) [reply pkg/lib/message.go:52]: Yea, if we can.

**Comment by mfdeveloper508** [comment]
Thanks guys
As you mentioned, I am going to add update codes, and some encoding/decoding logic for several data types, validation logic too.

In this PR, I think that main logic have been implemented.
- create dynamic data type form configuration
- encoding/decoding logic by category of types
- customize marshall/unmarshall of xml/json, 

new things are

- dynamic type
New approach is to use element struct instead of all of the data types.
https://github.com/moov-io/iso8583/pull/20/files#diff-22ddc508355be3ca89060ba288dde467b69ffc2953dcb58b18fa872a8ab0f36bR20

- encoding definition
The definition of encoding format isn't in ISO 8583
so I added a method what is to use encoding format.
https://github.com/moov-io/iso8583/pull/20/files#diff-af09e541e3096bb909d933dc53b25d2034af326f1a794cdc929d36aae7d2876dR65

Can you review the above new approaches first?
I want to take new idea through reviewing the PR.

**Comment by adamdecaf** [comment]
Is the idea with `Element` to embed it into other structs? Those structs can extend and offer their own methods specific to their behavior / data?

File: pkg/utils/version1987.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+			95:  {"an 42", "Replacement amounts"},
+			96:  {"b 64", "Message security code"},
+			97:  {"x+n 16", "Net settlement amount"},
+			98:  {"ans 25", "Payee"},
+			99:  {"n..11", "Settlement institution identification code"},
+			100: {"n..11", "Receiving institution identification code"},
+			101: {"ans..17", "File name"},
+			102: {"ans..28", "Account identification 1"},
+			103: {"ans..28", "Account identification 2"},
+			104: {"ans...100	", "Transaction description"},
```
Comment (SharonAllsup) [inline pkg/utils/version1987.go:122]: Need to remove white space between the length 100 and terminating quote

File: pkg/utils/version1987.go [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+			13:  {"n 4; MMDD", "Local transaction date (MMDD)"},
+			14:  {"n 4; YYMM", "Expiration date"},
+			15:  {"n 4; MMDD", "Settlement date"},
+			16:  {"n 4; MMDD", "Currency conversion date"},
+			17:  {"n 4; MMDD", "Capture date"},
+			18:  {"n 4", "Merchant type, or merchant category code"},
+			19:  {"n 3", "Acquiring institution (country code)"},
+			20:  {"n 3", "PAN extended (country code)"},
+			21:  {"n 3", "Forwarding institution (country code)"},
+			22:  {"n 3", "Point of service entry mode"},
```
Comment (SharonAllsup) [inline pkg/utils/version1987.go:40]: "Point of Sale (POS)", not "Point of service"

File: pkg/utils/version1987.go [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+			23:  {"n 3", "Application PAN sequence number"},
+			24:  {"n 3", "Function code (ISO 8583:1993), or network international identifier (NII)"},
+			25:  {"n 2", "Point of service condition code"},
```
Comment (SharonAllsup) [inline pkg/utils/version1987.go:43]: "Point of Sale (POS)", not "Point of service"

File: pkg/utils/version1987.go [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+			26:  {"n 2", "Point of service capture code"},
```
Comment (SharonAllsup) [inline pkg/utils/version1987.go:44]: "Point of Sale (POS)", not "Point of service" here, too

File: pkg/utils/version1987.go [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+			27:  {"n 1", "Authorizing identification response length"},
+			28:  {"x+n 8", "Amount, transaction fee"},
+			29:  {"x+n 8", "Amount, settlement fee"},
+			30:  {"x+n 8", "Amount, transaction processing fee"},
+			31:  {"x+n 8", "Amount, settlement processing fee"},
+			32:  {"n..11", "Acquiring institution identification code"},
+			33:  {"n..11", "Forwarding institution identification code"},
+			34:  {"ns..28", "Primary account number, extended"},
+			35:  {"z..37", "Track 2 data"},
+			36:  {"n...104", "Track 3 data"},
```
Comment (SharonAllsup) [inline pkg/utils/version1987.go:54]: Should Track 3 Data be n..104 or z..104?

File: pkg/lib/elements.go [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+	Text     string   `xml:",chardata"`
+	Elements []struct {
+		Text   string `xml:",chardata"`
+		Number int    `xml:"Number,attr"`
+	} `xml:"Element"`
+}
+
+// create data elements of message with specification
+func NewDataElements(spec *utils.Specification) (*dataElements, error) {
+	if spec == nil && spec.Elements == nil || spec.Encoding == nil {
```
Comment (alovak) [inline pkg/lib/elements.go:29]:
```suggestion
	if spec == nil || spec.Elements == nil || spec.Encoding == nil {
```

File: pkg/lib/lib_test.go [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+	assert.Equal(t, orgJsonBuf, jsonBuf)
+	assert.Equal(t, orgXmlBuf, xmlBuf)
+}
+
+func TestIso8583Message(t *testing.T) {
+	jsonStr := []byte(`
+	{
+		"mti": "0800",
+		"bitmap": "10100000001000000000000000000000000001",
```
Comment (alovak) [inline pkg/lib/lib_test.go:60]: I would expect a bitmap to be generated based on provided elements, not provided as input together with elements. Am I missing something?

File: pkg/lib/element.go [hunk 16]
```diff
@@ -0,0 +1,331 @@
...
+		hexStr := fmt.Sprintf("%0"+strconv.Itoa(e.Length/4)+"s", strconv.FormatUint(bitNum, 16))
+		value = []byte(hexStr)
+	} else {
+		return nil, errors.New(utils.ErrInvalidEncoder)
+	}
+
+	return value, nil
+}
+
+func (e *Element) characterDecoding(raw []byte) (int, error) {
```
[the remark this answers is not in the document]
Comment (mfdeveloper508) [reply pkg/lib/element.go:210]: yes, we need to add error check for this.

File: pkg/lib/element.go [hunk 17]
```diff
@@ -0,0 +1,331 @@
...
+	if !e.Fixed {
+		lenSize := len(strconv.Itoa(e.Length))
+		bcdSize := lenSize / 2
+		if lenSize%2 != 0 {
+			bcdSize++
+		}
+
+		switch e.LengthEncoding {
+		case utils.EncodingAscii:
+			contentLen, err = strconv.Atoi(string(raw[:lenSize]))
```
[the remark this answers is not in the document]
Comment (mfdeveloper508) [reply pkg/lib/element.go:225]: yes, we need to add error check for this.

File: pkg/lib/element.go [hunk 18]
```diff
@@ -0,0 +1,331 @@
...
+	}
+
+	if contentLen == 0 {
+		contentLen = e.Length
+	} else {
+		e.DataLength = contentLen
+	}
+
+	if e.Encoding == utils.EncodingAscii {
+		value, err = utils.UTF8ToWindows1252(raw[read : read+contentLen])
```
[the remark this answers is not in the document]
Comment (mfdeveloper508) [reply pkg/lib/element.go:262]: yes, we need to add error check for this.

File: pkg/lib/lib_test.go (hunk 15 above)
Comment (mfdeveloper508) [reply pkg/lib/lib_test.go:60]: added TestIso8583MessageBytes

**Comment by mfdeveloper508** [comment]
> Is the idea with `Element` to embed it into other structs? Those structs can extend and offer their own methods specific to their behavior / data?

Element implemented special functions according to data type,  data encoding, length encoding like as Load, Bytes functions.
I think that we can reduce many data structs using general Element struct.

File: pkg/utils/bcd.go (hunk 3 above)
Comment (mfdeveloper508) [reply pkg/utils/bcd.go]: The part is used to specify blank field of BCD encoding/decoding.
so default is 0x0 (the value is not the value that used in the bcd library)

File: go.mod (hunk 8 above)
[the remark this answers is not in the document]
Comment (mfdeveloper508) [reply go.mod:6]: replaced with https://github.com/indece-official/go-ebcdic (MIT License)

File: pkg/utils/types.go [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+				if len(splits) > 1 {
+					format = strings.TrimSpace(splits[len(splits)-1])
+				}
+			}
+			splits := strings.Split(attribute, indicate)
+			if len(splits) > 1 {
+				isFixed := !strings.Contains(indicate, variableIndicate)
+				_size, err := strconv.Atoi(strings.TrimSpace(splits[len(splits)-1]))
+				if err != nil ||
+					(!isFixed && _size > int(math.Pow(10, float64(len(indicate))))) {
```
Comment (adamdecaf) [inline pkg/utils/types.go:50]: Do we need to split this on a second line? It surprised me at first.

**Review by SharonAllsup — APPROVED** [review]

**Review by SharonAllsup — APPROVED** [review]

**Comment by mfdeveloper508** [comment]
Merge Comment:
add new library and utility to manage ISO message with configuration file

**Review by adamdecaf — CHANGES_REQUESTED** [review]
Can we fix the LGTM error?

**Comment by mfdeveloper508** [comment]
fixed the LGTM error

**Review by adamdecaf — APPROVED** [review]

**Comment by adamdecaf** [comment]
The code looks good. We got some easy fixes out of the way when building the PR. I'm not super familiar with the low level details, but the code is fixable and adaptable as we continue forward. 

It would be nice to add a fuzzer or two for the low level parsers. Those will need to be rock solid. We've got these in other projects that are ran continiously

https://github.com/moov-io/ach/tree/master/test/fuzz-reader

**Comment by mfdeveloper508** [comment]
yep, I will fix fuzz-reader and another issues of github too

**Comment by adamdecaf** [comment]
Sounds good. Can we merge this?

**Comment by mfdeveloper508** [comment]
sure, 
After merge, I will use another PR for special issues

**Comment by adamdecaf** [comment]
merged, see https://github.com/moov-io/iso8583/commit/7a97123561026f8737fb8071ad2698eeb0f5da12

**Pull request closed without merging**